stats/lib/platform-stats-fetchers.js

```diff
@@ -43,6 +43,22 @@ export const fetchDailyStationAcceptedMeasurementCount = async (pgPool, filter)
   return rows
 }
 
+/**
+ * @param {Queryable} pgPool
+ * @param {import('./typings.js').DateRangeFilter} filter
+ */
+export const fetchTopMeasurementStations = async (pgPool, filter) => {
+  const { rows } = await pgPool.query(`
+    SELECT station_id, SUM(accepted_measurement_count) as total_accepted_measurement_count
+    FROM daily_stations
+    WHERE day >= $1 AND day <= $2
+    GROUP BY station_id
+    ORDER BY total_accepted_measurement_count DESC
```

Comment:
Do you know how expensive this query will be? Are there any indexes we can add to make it run faster?

Reply:
I mention this in my above comment: https://github.com/filecoin-station/spark-stats/pull/164#issuecomment-2186798998

It doesn't look like an index will be perfect, but supposedly still helpful.

Reply:
Well, it is expensive.

```
# EXPLAIN ANALYZE SELECT station_id, SUM(accepted_measurement_count) as total_accepted_measurement_count
FROM daily_stations
WHERE day >= (now() - INTERVAL '7 day') AND day <= now()
GROUP BY station_id
ORDER BY total_accepted_measurement_count DESC
;
                                                                QUERY PLAN                                                                
------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=856016.50..860259.08 rows=1697033 width=97) (actual time=6236.102..6507.571 rows=1854497 loops=1)
   Sort Key: (sum(accepted_measurement_count)) DESC
   Sort Method: external merge  Disk: 206928kB
   ->  HashAggregate  (cost=495180.10..571699.85 rows=1697033 width=97) (actual time=4152.936..5149.416 rows=1854497 loops=1)
         Group Key: station_id
         Planned Partitions: 128  Batches: 129  Memory Usage: 4369kB  Disk Usage: 253808kB
         ->  Seq Scan on daily_stations  (cost=0.00..378860.25 rows=2032620 width=93) (actual time=83.001..2948.713 rows=2128699 loops=1)
               Filter: ((day <= now()) AND (day >= (now() - '7 days'::interval)))
               Rows Removed by Filter: 7102107
 Planning Time: 0.105 ms
 JIT:
   Functions: 12
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 1.820 ms, Inlining 15.963 ms, Optimization 70.940 ms, Emission 41.194 ms, Total 129.916 ms
 Execution Time: 6655.125 ms
```

Reply:
> I mention this in my above comment: [#164 (comment)](https://github.com/filecoin-station/spark-stats/pull/164#issuecomment-2186798998)
> 
> It doesn't look like an index will be perfect, but supposedly still helpful.

Ah, I missed your comment. Thank you for the link!

Reply:
I have tried this myself with some test data, and mine is much faster, and I'm not sure why... Roughly 850ms. And adding an index as I described above seems to have no effect

```
import http from 'node:http'
import { once } from 'node:events'
import assert from 'node:assert'
import createDebug from 'debug'
import { getPgPools } from '@filecoin-station/spark-stats-db'

import { assertResponseStatus, getPort } from '../test/test-helpers.js'
import { createHandler } from '../lib/handler.js'

const debug = createDebug('test')

const givenDailyStationMetrics = async (pgPoolEvaluate, day, stationStats) => {
    await pgPoolEvaluate.query(`
      INSERT INTO daily_stations (day, station_id, accepted_measurement_count)
      SELECT $1 AS day, UNNEST($2::text[]) AS station_id, UNNEST($3::int[]) AS accepted_measurement_count
      ON CONFLICT DO NOTHING
      `, [
      day,
      stationStats.map(s => s.stationId),
      stationStats.map(s => s.acceptedMeasurementCount)
    ])
  }

let pgPools = await getPgPools()
await pgPools.evaluate.query('DELETE FROM daily_stations')

const handler = createHandler({
pgPools,
logger: {
    info: debug,
    error: console.error,
    request: debug
}
})

let server = http.createServer(handler)
server.listen()
await once(server, 'listening')
const baseUrl = `http://127.0.0.1:${getPort(server)}`

for (let day = '2024-01-01'; day <= '2024-01-09'; day = day.substring(0, 8) + (parseInt(day.substring(8)) + 1).toString().padStart(2, '0')) {
  console.log(day)
  const station_stats = []
  for (let i = 0; i < 200000; i++) {
    // await givenDailyStationMetrics(pgPools.evaluate, day, [
      station_stats.push({ stationId: `station${i}`, acceptedMeasurementCount: Math.floor(Math.random() * 1000) })
    // ])
  }
  await givenDailyStationMetrics(pgPools.evaluate, day, station_stats)
}

// assert 1,800,000 records were inserted
const { rows: [{ count }] } = await pgPools.evaluate.query('SELECT COUNT(*) FROM daily_stations')
assert.strictEqual(count, '1800000')

const startTime = Date.now()

const res = await fetch(
  new URL(
    '/stations/top-measurements?from=2024-01-01&to=2024-01-02',
    baseUrl
  ), {
    redirect: 'manual'
  }
)

console.log(await res.json())

const executionTime = Date.now() - startTime
console.log('Execution time:', executionTime)
```

Reply:
Ah looks like one difference is your query did not have a LIMIT. But even if I remove it, mine is only 2300ms

Reply:
```
spark_evaluate=#     EXPLAIN ANALYZE SELECT station_id, SUM(accepted_measurement_count) as total_accepted_measurement_count
    FROM daily_stations
    WHERE day >= '2024-01-01' AND day <= '2024-01-09'
    GROUP BY station_id
    ORDER BY total_accepted_measurement_count DESC;
                                                                           QUERY PLAN                                                                            
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=210141.11..210639.40 rows=199316 width=21) (actual time=903.301..927.241 rows=200000 loops=1)
   Sort Key: (sum(accepted_measurement_count)) DESC
   Sort Method: external merge  Disk: 6672kB
   ->  Finalize GroupAggregate  (cost=138011.99..188508.62 rows=199316 width=21) (actual time=604.484..842.231 rows=200000 loops=1)
         Group Key: station_id
         ->  Gather Merge  (cost=138011.99..184522.30 rows=398632 width=21) (actual time=604.475..744.019 rows=579498 loops=1)
               Workers Planned: 2
               Workers Launched: 2
               ->  Sort  (cost=137011.96..137510.25 rows=199316 width=21) (actual time=595.138..625.003 rows=193166 loops=3)
                     Sort Key: station_id
                     Sort Method: external merge  Disk: 6512kB
                     Worker 0:  Sort Method: external merge  Disk: 6464kB
                     Worker 1:  Sort Method: external merge  Disk: 6344kB
                     ->  Partial HashAggregate  (cost=104597.25..115379.47 rows=199316 width=21) (actual time=334.234..453.020 rows=193166 loops=3)
                           Group Key: station_id
                           Planned Partitions: 4  Batches: 5  Memory Usage: 8241kB  Disk Usage: 15776kB
                           Worker 0:  Batches: 5  Memory Usage: 8241kB  Disk Usage: 15376kB
                           Worker 1:  Batches: 5  Memory Usage: 8241kB  Disk Usage: 15704kB
                           ->  Parallel Seq Scan on daily_stations  (cost=0.00..50691.00 rows=750000 width=17) (actual time=11.587..133.580 rows=600000 loops=3)
                                 Filter: ((day >= '2024-01-01'::date) AND (day <= '2024-01-09'::date))
 Planning Time: 0.225 ms
 Execution Time: 945.630 ms
(22 rows)

```

Reply:
Performance analysis of the new query - `Execution Time: 1924.542 ms`

```
spark_evaluate=# EXPLAIN ANALYZE SELECT
    participant_address,
    COUNT(DISTINCT inet_group) AS inet_group_count,
    COUNT(DISTINCT station_id) AS station_count,
    SUM(accepted_measurement_count) AS accepted_measurement_count
  FROM daily_stations
  WHERE day >= DATE(now()) AND day <= DATE(now())
  GROUP BY participant_address
  ORDER BY accepted_measurement_count DESC;
                                                                               QUERY PLAN                                           
                                    
------------------------------------------------------------------------------------------------------------------------------------
------------------------------------
 Sort  (cost=103736.43..103736.93 rows=200 width=56) (actual time=1912.549..1912.978 rows=3071 loops=1)
   Sort Key: (sum(accepted_measurement_count)) DESC
   Sort Method: quicksort  Memory: 528kB
   ->  GroupAggregate  (cost=101834.29..103728.79 rows=200 width=56) (actual time=963.668..1911.692 rows=3071 loops=1)
         Group Key: participant_address
         ->  Sort  (cost=101834.29..102212.79 rows=151400 width=157) (actual time=963.616..1024.363 rows=333525 loops=1)
               Sort Key: participant_address
               Sort Method: external merge  Disk: 50496kB
               ->  Index Scan using daily_stations_pkey on daily_stations  (cost=0.69..81835.78 rows=151400 width=157) (actual time=
7.956..723.050 rows=333525 loops=1)
                     Index Cond: ((day >= date(now())) AND (day <= date(now())))
 Planning Time: 0.156 ms
 JIT:
   Functions: 11
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 1.683 ms, Inlining 0.000 ms, Optimization 0.447 ms, Emission 7.201 ms, Total 9.331 ms
 Execution Time: 1924.542 ms
(16 rows)
```

Reply:
Experiment: 

```sql
CREATE INDEX mbtest ON daily_stations (day, participant_address);
```

Performance - `Execution Time: 1536.079 ms` - a very small improvement.

```
 Sort  (cost=96047.44..96047.94 rows=200 width=56) (actual time=1526.023..1526.609 rows=3071 loops=1)
   Sort Key: (sum(accepted_measurement_count)) DESC
   Sort Method: quicksort  Memory: 528kB
   ->  GroupAggregate  (cost=94153.83..96039.79 rows=200 width=56) (actual time=361.709..1525.309 rows=3071 loops=1)
         Group Key: participant_address
         ->  Sort  (cost=94153.83..94530.62 rows=150717 width=157) (actual time=361.682..446.334 rows=333525 loops=1)
               Sort Key: participant_address
               Sort Method: external merge  Disk: 50512kB
               ->  Index Scan using mbtest on daily_stations  (cost=0.45..74249.75 rows=150717 width=157) (actual time=0.038..176.38
4 rows=333525 loops=1)
                     Index Cond: ((day >= date(now())) AND (day <= date(now())))
 Planning Time: 0.219 ms
 Execution Time: 1536.079 ms
```

Reply:
Experiment 2:

```sql
```

Performance - `Execution Time: 2416.290 ms` 👎🏻 

```
 Sort  (cost=104484.10..104484.60 rows=200 width=56) (actual time=2403.051..2403.494 rows=3093 loops=1)
   Sort Key: (sum(accepted_measurement_count)) DESC
   Sort Method: quicksort  Memory: 531kB
   ->  GroupAggregate  (cost=102570.31..104476.46 rows=200 width=56) (actual time=1273.335..2402.412 rows=3093 loops=1)
         Group Key: participant_address
         ->  Sort  (cost=102570.31..102951.14 rows=152332 width=157) (actual time=1273.285..1352.646 rows=447542 loops=1)
               Sort Key: participant_address
               Sort Method: external merge  Disk: 68584kB
               ->  Index Scan using daily_stations_pkey on daily_stations  (cost=0.69..82441.81 rows=152332 width=157) (actual time=
6.874..966.039 rows=447542 loops=1)
                     Index Cond: ((day >= date(now())) AND (day <= date(now())))
 Planning Time: 0.095 ms
 JIT:
   Functions: 11
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 1.355 ms, Inlining 0.000 ms, Optimization 0.332 ms, Emission 6.294 ms, Total 7.981 ms
 Execution Time: 2416.290 ms
```

Reply:
Measuring the performance with no index again - `Execution Time: 2546.602 ms`.

I think as we are recording more data, the query becomes even more expensive when there is no index.

```
 Sort  (cost=104484.10..104484.60 rows=200 width=56) (actual time=2531.676..2532.188 rows=3093 loops=1)
   Sort Key: (sum(accepted_measurement_count)) DESC
   Sort Method: quicksort  Memory: 531kB
   ->  GroupAggregate  (cost=102570.31..104476.46 rows=200 width=56) (actual time=1366.055..2531.003 rows=3093 loops=1)
         Group Key: participant_address
         ->  Sort  (cost=102570.31..102951.14 rows=152332 width=157) (actual time=1366.008..1446.075 rows=447542 loops=1)
               Sort Key: participant_address
               Sort Method: external merge  Disk: 68584kB
               ->  Index Scan using daily_stations_pkey on daily_stations  (cost=0.69..82441.81 rows=152332 width=157) (actual time=
7.966..1031.773 rows=447542 loops=1)
                     Index Cond: ((day >= date(now())) AND (day <= date(now())))
 Planning Time: 0.155 ms
 JIT:
   Functions: 11
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 1.583 ms, Inlining 0.000 ms, Optimization 0.425 ms, Emission 7.309 ms, Total 9.317 ms
 Execution Time: 2546.602 ms
```

Reply:
Let's see how long this query will take on Monday.

I suspect we will need to add an index `ON daily_stations (day, participant_address)`

@PatrickNercessian What other options do we have to make this query run faster? You mentioned materialized views, how would that work?

Reply:
This will not necessarily make it faster, but moreso invisible to the any user. The real issue with any slow query today is that the first cache miss will take a long time.

A Materialized View (as I understand it), is basically a virtual table that is constructed by a normal query on other tables on a schedule. Then our actual API queries are basically instant, since they're just reading from the table (e.g. `SELECT * FROM top_measurement_stations_mv`)

For example,

```
CREATE MATERIALIZED VIEW top_measurement_stations_mv AS
SELECT
  participant_address,
  COUNT(DISTINCT inet_group) AS inet_group_count,
  COUNT(DISTINCT station_id) AS station_count,
  SUM(accepted_measurement_count) AS accepted_measurement_count
FROM daily_stations
WHERE day = CURRENT_DATE - INTERVAL '1 day'
GROUP BY participant_address
ORDER BY accepted_measurement_count DESC;
```

This may not be perfect, but the idea above works. Then we can run 

`REFRESH MATERIALIZED VIEW top_measurement_stations_mv;`

on a schedule, such as once per day.

Reply:
I am running the query now to fetch stats for yesterday.

```
spark_evaluate=# EXPLAIN ANALYZE SELECT
  participant_address,
  COUNT(DISTINCT inet_group) AS inet_group_count,
  COUNT(DISTINCT station_id) AS station_count,
  SUM(accepted_measurement_count) AS accepted_measurement_count
FROM daily_stations
WHERE day >= DATE(now() - INTERVAL '1 day') AND day <= DATE(now() - INTERVAL '1 day')
GROUP BY participant_address
ORDER BY accepted_measurement_count DESC;
                                                                            QUERY PLAN                                                                            
------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=2311999.04..2311999.12 rows=32 width=59) (actual time=33629.984..33689.545 rows=3408 loops=1)
   Sort Key: (sum(accepted_measurement_count)) DESC
   Sort Method: quicksort  Memory: 576kB
   ->  GroupAggregate  (cost=1512818.68..2311998.24 rows=32 width=59) (actual time=6176.119..33688.200 rows=3408 loops=1)
         Group Key: participant_address
         ->  Gather Merge  (cost=1512818.68..2248804.95 rows=6319297 width=144) (actual time=6176.076..8616.084 rows=6274167 loops=1)
               Workers Planned: 2
               Workers Launched: 2
               ->  Sort  (cost=1511818.66..1518401.26 rows=2633040 width=144) (actual time=6064.667..6481.303 rows=2091389 loops=3)
                     Sort Key: participant_address
                     Sort Method: external merge  Disk: 329000kB
                     Worker 0:  Sort Method: external merge  Disk: 326776kB
                     Worker 1:  Sort Method: external merge  Disk: 338912kB
                     ->  Parallel Seq Scan on daily_stations  (cost=0.00..1009635.55 rows=2633040 width=144) (actual time=197.807..4200.450 rows=2091389 loops=3)
                           Filter: ((day >= date((now() - '1 day'::interval))) AND (day <= date((now() - '1 day'::interval))))
                           Rows Removed by Filter: 8065788
 Planning Time: 0.181 ms
 JIT:
   Functions: 15
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 3.134 ms, Inlining 173.461 ms, Optimization 171.399 ms, Emission 118.043 ms, Total 466.037 ms
 Execution Time: 33731.846 ms
(22 rows)
```

Reply:
With index `ON daily_stations (day, participant_address)`:

```
 Planning Time: 0.222 ms
 JIT:
   Functions: 15
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 3.101 ms, Inlining 161.029 ms, Optimization 168.619 ms, Emission 117.287 ms, Total 450.036 ms
 Execution Time: 38431.564 ms
```

Notice that Postgres is not able to use an index for the initial filtering by date and always runs a Parallel Seq Scan 😞

Reply:
FWIW, there were 3,408 active participants yesterday.

As I understand the output of `EXPLAIN ANALYZE`, the most expensive operation is sorting the results according to `accepted_measurement_count` - it takes about 30 seconds from the total query time of 36 seconds.

Reply:
Agreed, a 30s query is really heavy on the backend and creates a bad experience when uncached